### PR TITLE
add integration tests

### PR DESCRIPTION
**Pull request by kaisecheng**
The PR does integration tests against two SNMP dockers with ipv6 to simulate the scenarios below.
- mix udp tcp
- multiple hosts
- snmpv3 connection
- invalid user
- multiple inputs with mix tcp udp hosts
- same security name mix of udp tcp hosts
- ipv6 connection

### CODE REVIEW

File: spec/inputs/integration/it_spec.rb [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+describe LogStash::Inputs::Snmp do
+  let(:config) { {"get" => ["1.3.6.1.2.1.1.1.0", "1.3.6.1.2.1.1.3.0", "1.3.6.1.2.1.1.5.0"]} }
+  let(:plugin) { LogStash::Inputs::Snmp.new(config)}
+
+  after(:each) do
+    plugin.stop
+  end
+
+  shared_examples "snmp plugin return single event" do
+    it "should has OID value" do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:13]:
```suggestion
    it "should have OID value" do
```

File: spec/inputs/integration/it_spec.rb [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+      expect(event).to be_a(LogStash::Event)
+      expect(event.get("iso.org.dod.internet.mgmt.mib-2.system.sysUpTime.sysUpTimeInstance")).to be_a Integer
+      expect(event.get("iso.org.dod.internet.mgmt.mib-2.system.sysName.0")).to be_a String
+      expect(event.get("iso.org.dod.internet.mgmt.mib-2.system.sysDescr.0")).to be_a String
+    end
+  end
+
+  shared_examples "snmp plugin return one udp event and one tcp event" do |config|
+    it "should has one udp from snmp1 and one tcp from snmp2" do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:28]:
```suggestion
    it "should have one udp from snmp1 and one tcp from snmp2" do
```

File: spec/inputs/integration/it_spec.rb [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                                   "hosts" => [{"host" => "tcp:snmp1/161", "version" => "3"}],
+                                   "security_name" => "user_2",
+                                   "auth_protocol" => "sha",
+                                   "auth_pass" => "STrP@SSPhr@sE",
+                                   "priv_protocol" => "aes",
+                                   "priv_pass" => "STr0ngP@SSWRD161",
+                                   "security_level" => "authPriv"
+                               })}
+
+    it "should has error log" do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:79]:
```suggestion
    it "should have error log" do
```

File: spec/inputs/integration/it_spec.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        stop_plugin_after_seconds(plugin)
+        plugin.run(queue)
+      }
+      t2 = Thread.new {
+        stop_plugin_after_seconds(plugin2)
+        plugin2.run(queue)
+      }
+      t.join(2100)
+      t2.join(2100)
+      plugin2.close
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:136]: did we also want to `plugin.close` ?

**Comment by colinsurprenant** [comment]
Great stuff @kaisecheng! I added a few minor/cosmetic comments. I will be spending more time to run the new testing and followup shortly.

File: spec/inputs/integration/it_spec.rb (hunk 4 above)
Comment (kaisecheng) [reply spec/inputs/integration/it_spec.rb:137]: yes. we want. I was having `after(:each) plugin.stop`, but `close` is the right one. So, this need to fix

**Comment by colinsurprenant** [comment]
I believe https://github.com/logstash-plugins/.ci/pull/16 will solve the build problem on 8.0

File: spec/inputs/integration/it_spec.rb [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+          expect(event.get("[@metadata][host_address]")).to eq("snmp2")
+          expect(event.get("[@metadata][host_port]")).to eq("162")
+        end
+      }
+      expect(udp).to eq(1)
+      expect(tcp).to eq(1)
+    end
+  end
+
+  describe "against snmp server", :integration => true do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:46]:
```suggestion
  describe "against single snmp server with snmpv2 and udp", :integration => true do
```

File: spec/inputs/integration/it_spec.rb [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    let(:config) { super.merge({"hosts" => [{"host" => "udp:snmp1/161", "community" => "public"}]})}
+    it_behaves_like "snmp plugin return single event"
+  end
+
+  describe "against snmpv3 server", :integration => true do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:51]:
```suggestion
  describe "against single server with snmpv3 and tcp", :integration => true do
```

File: spec/inputs/integration/it_spec.rb [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+       "auth_pass" => "STrP@SSPhr@sE",
+       "priv_protocol" => "aes",
+       "priv_pass" => "STr0ngP@SSWRD161",
+       "security_level" => "authPriv"
+                               })}
+
+    it_behaves_like "snmp plugin return single event"
+  end
+
+  describe "invalid user against snmpv3 server", :integration => true do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:65]:
```suggestion
  describe "invalid user against single server with snmpv3 and tcp", :integration => true do
```

File: spec/inputs/integration/it_spec.rb [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      expect(plugin.logger).to receive(:error).once
+      plugin.register
+      queue = []
+      stop_plugin_after_seconds(plugin)
+      plugin.run(queue)
+      plugin.close
+    end
+  end
+
+  describe "single plugin input and mix of udp tcp", :integration => true do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:86]:
```suggestion
  describe "single input plugin on single server with snmpv2 and mix of udp and tcp", :integration => true do
```

File: spec/inputs/integration/it_spec.rb [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    let(:config) { super.merge({"hosts" => [{"host" => "udp:snmp1/161", "community" => "public"}, {"host" => "tcp:snmp1/161", "community" => "public"}]})}
+    it "should return two events " do
+      plugin.register
+      queue = []
+      stop_plugin_after_seconds(plugin)
+      plugin.run(queue)
+      plugin.close
+
+      host_cnt_snmp1 = queue.reduce(0) { |sum, event| sum + (event.get("host") == "snmp1"? 1: 0) }
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:95]: would something like this be simpler and/or easier to understand?
```ruby
host_cnt_snmp1 = queue.select {|event| event.get("host") == "snmp1"}.size
```

File: spec/inputs/integration/it_spec.rb [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      expect(queue.size).to eq(2)
+      expect(host_cnt_snmp1).to eq(2)
+    end
+  end
+
+  describe "single plugin input and multiple udp hosts", :integration => true do
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:101]:
```suggestion
  describe "single input plugin on multiple udp hosts", :integration => true do
```

File: spec/inputs/integration/it_spec.rb [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    let(:config) { super.merge({"hosts" => [{"host" => "udp:snmp1/161", "community" => "public"}, {"host" => "udp:snmp2/162", "community" => "public"}]})}
+    it "should return two events, one per host" do
+      plugin.register
+      queue = []
+      stop_plugin_after_seconds(plugin)
+      plugin.run(queue)
+      plugin.close
+
+      hosts = queue.map { |event| event.get("host") }.sort
```
Comment (colinsurprenant) [inline spec/inputs/integration/it_spec.rb:110]: or this method here could be used instead of my suggestion with `select` for the sake of consitency?

**Comment by colinsurprenant** [comment]
@kaisecheng again, this is great work, overall LGTM, left a few minor comments. Once these are addressed and this is rebased we should be good to merge!

**Pull request merged**